crates/resolver/src/name_server/name_server.rs

```diff
@@ -254,6 +259,21 @@ where
     NameServer::new_with_provider(config, options, conn_provider)
 }
 
+/// Used for connection mocking
+pub trait CreateConnection: Sized {
+    /// Future of Self
+    type FutureConn<P: RuntimeProvider>: Future<Output = Result<Self, ResolveError>>
+        + Send
+        + 'static;
+
+    /// Create a future of Self with the help of runtime provider.
+    fn new_connection<P: RuntimeProvider>(
+        runtime_provider: &P,
+        config: &NameServerConfig,
+        options: &ResolverOpts,
+    ) -> Self::FutureConn<P>;
+}
```

Comment:
Why is this necessary. If it's just for mocking, can it be scoped to a `#[cfg(test)]` block?

Reply:
Because I introduce this trait as an intermediate layer. If we didn't need testing, we could remove this redundant trait and move the code to actual implementation. Therefore, it can not be labeled as `test`. I will modify this confusing comment.

Reply:
Why do we need this intermediate layer, though?